CASSANDRA.md

```diff
@@ -0,0 +1,543 @@
+    CacheBuilder.newBuilder().
+      maximumSize(nrOfCacheEntries).
+      build(
+        new CacheLoader[String, PreparedStatement]() {
+          def load(key: String): PreparedStatement = session.prepare(key.toString)
+        }
+      )
+
+  case class WeatherStation(country: String, city: String, stationId: String, entry: Int, value: Int)
+
+  object WeatherStation {
+
+    def getAllByCountry(cache: LoadingCache[String, PreparedStatement], session: Session)(country: String): List[WeatherStation] = {
+      val query: Statement =
+        QueryBuilder.select().
+          all().
+          from("db", "weather_station").
+          where(QueryBuilder.eq("country", QueryBuilder.bindMarker()))
+
+      session.execute(cache.get(query.toString).bind(country)).all().asScala.map(
+        row => WeatherStation(row.getString("country"), row.getString("city"), row.getString("station_id"), row.getInt("entry"), row.getInt("value"))
+      ).to[List]
+    }
+  }
+
+  val getAllByCountry: String => List[WeatherStation] = WeatherStation.getAllByCountry(cache, session)_
+
+  getAllByCountry("UK")
+
+  session.close()
+  cluster.close()
+}
+```
+
+The Java driver requires explicit handling of a `PreparedStatement`s cache to avoid preparing the same statement more that once, that could affect performance.
+
+**Phantom (v1.22.0)**
+```scala
+import com.websudos.phantom.connectors.RootConnector
+import com.websudos.phantom.db._
+import com.websudos.phantom.dsl._
+
+import scala.concurrent.Future
+
+object Phantom extends App {
+
+  case class WeatherStation(country: String, city: String, stationId: String, entry: Int, value: Int)
+
+  abstract class WeatherStationCF(override val tableName: String) extends CassandraTable[WeatherStationCF, WeatherStation] with RootConnector {
+
+    object country extends StringColumn(this) with PartitionKey[String]
+    object city extends StringColumn(this) with PrimaryKey[String]
+    object stationId extends StringColumn(this) with PrimaryKey[String] {
+      override lazy val name: String = "station_id"
+    }
+    object entry extends IntColumn(this) with PrimaryKey[Int]
+    object value extends IntColumn(this)
+
+    override def fromRow(r: Row): WeatherStation =
+      WeatherStation(country(r), city(r), stationId(r), entry(r), value(r))
+  }
+
+  abstract class WeatherStationQueries extends WeatherStationCF("weather_station") {
+
+    def getAllByCountry(country: String): Future[List[WeatherStation]] =
+      select.where(_.country eqs country).fetch()
+  }
+
+  class DB(ks: KeySpaceDef) extends DatabaseImpl(ks) {
+
+    object stations extends WeatherStationQueries with connector.Connector
+  }
+
+  val db = new DB(ContactPoint.local.keySpace("db"))
+
+  val result = db.stations.getAllByCountry("UK")
+
+  result.onComplete(_ => db.shutdown())
+}
+```
+
+Phantom requires explicit definition of the database model. The query definition also requires special equality operators.
```

Comment:
"Phantom requires explicit definition of the database model"
=>
"Phantom requires mapping classes to lift the database model to DSL types"

wdyt?